### PR TITLE
[InstCombine] Optimistically allow multiple shufflevector uses in foldOpPhi

### PR DESCRIPTION
**Pull request by MatzeB**
We would like to optimize situations of the form that happen after loop vectorization+SROA:
```
loop:
    %phi = phi zeroinitializer, %interleaved

    %deinterleave_a = shufflevector %phi, poison ; pick half of the lanes
    %deinterleave_b = shufflevector %phi, posion ; pick remaining lanes

    ... %a = ... %b = ...

    %interleaved = shufflevector %a, %b ; interleave lanes of a+b
```
where the interleave and de-interleave shuffle operations cancel each other out.
This could be handled by `foldOpPhi` but does not currently work because it does
not proceed when there are multiple uses of the `Phi` operation.

This extends `foldOpPhi` to allow multiple `shufflevector` uses when they are
shown to simplify for all `Phi` input values.

### CODE REVIEW

**Comment by MatzeB** [comment]
This pattern happens when autovectorizing an important inner loop of fbgemm  ( https://github.com/pytorch/FBGEMM/blob/119052f4dddc08064ea32d058c05055f1a25ec74/src/EmbeddingSpMDMAutovec.cc#L418 ); a simplified version of this is shown in the new `test/Transforms/InstCombine/vec_shuffle-phi-multiuse.ll`.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1786]:
```suggestion
    MultipleUses = isa<ShuffleVectorInst>(I);
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1778]: `MultipleUses` is misleading. `MultipleShuffleVectorUses` would be better.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1919]: Can you explain this change?

Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1919]: When there is multipluses operands using the PHI, i.e. `op0(Phi(...))` and `op1(Phi(...))` then this loop will replace all uses of the `Phi` with the new `Phi(op0(...)...)` pattern, this includes replacing `op1` so this is only a correct transformation when op0 and op1 are identical.

The shufflevector code I add here `op0` and `op1` may be shufflevectors with different masks so we have can only optimize one use at a time; but instcombine will invoke this rule a 2nd for the other use later so it will get optimized as well.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1778]: In principle we can add more exceptions in the future to allow multiple uses in more situations, so I used the more generic name. Anyway will rename to `MultipleShuffleVectorUses` then as we can always rename in the future when someone actually allows more cases...

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Review by nikic — COMMENTED** [review]
If I'm reading this correctly, then nothing here actually guarantees that *all* shufflevectors are going to simplify? You just check that all uses are shufflevectors, but then simplify each individually.

If we say that we consider it profitable to replace a shufflevector with a phi in general, then we shouldn't need the check on the uses for shufflevectors at all.

**Comment by nikic** [comment]
> > If I'm reading this correctly, then nothing here actually guarantees that all shufflevectors are going to simplify? You just check that all uses are shufflevectors, but then simplify each individually.
> 
> See L1851-1855:
> 
> ```
>      // Be conservative in cases with multiple uses and require all inputs to
>      // simplify.
> 
>      if (MultipleShuffleVectorUses)
>        return nullptr;
> ```

Right, but doesn't think just check that *this* shufflevector simplifies (for all phi inputs), not that *all* shufflevectors simplify?

File: llvm/test/Transforms/InstCombine/vec_shuffle-phi-multiuse.ll
Comment (dmpots) [inline llvm/test/Transforms/InstCombine/vec_shuffle-phi-multiuse.ll:4]: Is it worth adding tests for cases where it should not simplify (e.g. too many users, some users can't simplify)?

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dmpots) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1796]: Is there something special about `shufflevector` that we would only want this to apply to that type of instruction? It seems like the other checks are generic and as long as the users all simplify it should be ok?

**Review by dmpots — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1796]: Generally speaking you have to be careful when optimizing instcombine patterns when any node (but the root node) in the pattern has multiple uses. It's generally not easy to confirm that everything does indeed simplify and you can easily turn one value into multiple values being alive increase register pressure and ending up with worse performance. Unfortunately I don't think we have good ways to predict/model this in LLVm so you will notice many instcombine being conservative and bailing out on multiple uses.

My reasoning to make an exception for this particular case is that this happens to hit a very important kernel in our workload; and "strided vectorization" being an explicit mode in the loop vectorizer that can produce patterns like my example "by design" (after SROA).

Comment (dmpots) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1796]: I see. I thought the check [below](https://github.com/llvm/llvm-project/pull/114278/files#diff-0ffb024c6c026c4b93507d08e6d24dccf98980023130781164b71586bc747f2dR1851) was to ensure that the shuffle vector uses were simplified.

```
    // Be conservative in cases with multiple uses and require all inputs to
    // simplify.
    if (MultipleShuffleVectorUses)
      return nullptr;
```

But it sounds like even with this check we may end up with some shufflevectors that are not simplified?

Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1796]: I am still working on the feedback from above.

File: llvm/test/Transforms/InstCombine/vec_shuffle-phi-multiuse.ll
Comment (MatzeB) [reply llvm/test/Transforms/InstCombine/vec_shuffle-phi-multiuse.ll]: added more tests

**Comment by MatzeB** [comment]
> Right, but doesn't think just check that _this_ shufflevector simplifies (for all phi inputs), not that _all_ shufflevectors simplify?

new versions checks that all phi users simplify.

**Comment by MatzeB** [comment]
> If we really want to check all shufflevectors, I think the transform should be further separated from the existing one and simplify all the shufflevectors at once. It makes little sense to check that all of them simplify and compute the replacement values, then discard all of that and replace a single one of them, and then repeat the whole process again for the next shufflevector.

@nikic 
- Would you rather see this logic in a `visitPHINode` function that would then check all the users and optimize them all at once. I was considering doing that, but wasn't sure if people would apreciate the change in style as the transform rules would no longer be within `visitShuffleVector` where people may intuitively look for them.
- Or are you saying that we should rather just drop all the checks and just do the simple thing and always optimize regardless of amount of uses. In that case, what sort of qualification/testing would convince you that we don't regress on average?

**Comment by nikic** [comment]
I'm saying we should try the latter first and if that fails go to the former.

**Comment by MatzeB** [comment]
Here's the simpler version just optimistically transforming `shufflevector(phi(shufflevector()))` ignoring number of uses on the phi.

I did `llvm-test-suite` builds with/without this change and it only affected codegen in 6 benchmarks[1]. I was not able to measure any performance difference above noise level and inspecting the assembly I saw a couple shuffles moving out of loops, a bunch of churn that I couldn't whether it's better/worse, but also did not see anything obviously worse.

So worth a try optimistically enabling this.

[1]
External/SPEC/CFP2006/447.dealII/447.dealII.test
External/SPEC/CFP2006/453.povray/453.povray.test
External/SPEC/CINT2006/464.h264ref/464.h264ref.test
MultiSource/Benchmarks/Bullet/bullet.test
MultiSource/Applications/JM/lencod/lencod.test
MultiSource/Benchmarks/DOE-ProxyApps-C++/CLAMR/CLAMR.test

**Review by nikic — COMMENTED** [review]
Sorry for being unclear, I do think that you should still keep the !SeenNonSimplifiedInVal check for the multi-use case. Without that check we could end up with *both* the original phis and extra new shufflevectors (rather than replacing shufflevectors with phis).

**Comment by MatzeB** [comment]
updated.

**Comment by MatzeB** [comment]
@nikic now good to go?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
